src/mono/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,168 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.CompilerServices
+{
+    internal struct Ephemeron
+    {
+        public object? key;
+        public object? value;
+    }
+}
+
+namespace System.Runtime
+{
+    //
+    // Instead of dependent handles, mono uses arrays of Ephemeron objects.
+    //
+    public struct DependentHandle : IDisposable
+    {
+        private Ephemeron[] data;
+
+        public DependentHandle(object? target, object? dependent)
+        {
+            data = new Ephemeron[1];
+            data[0].key = target;
+            data[0].value = dependent;
+            GC.register_ephemeron_array(data);
+        }
+
+        public bool IsAllocated => data is not null;
+
+        public object? Target
+        {
+            get => GetTarget();
+            set => SetTarget(value);
+        }
+
+        public object? Dependent
+        {
+            get => GetDependent();
+            set => SetDependent(value);
+        }
+
+        public (object? Target, object? Dependent) GetTargetAndDependent()
+        {
+            if (this.data is not Ephemeron[] data)
+            {
+                ThrowHelper.ThrowInvalidOperationException();
+
+                return default;
+            }
+
+            if (data[0].key == GC.EPHEMERON_TOMBSTONE)
+            {
+                return default;
+            }
+
+            return (data[0].key, data[0].value);
```

Comment:
Is it possible data[0] could transition to GC.EPHEMERON_TOMBSTONE between the time it's checked above and here?  If yes, it should probably be more like:
```C#
Ephemeron e = data[0];
return e.key != GC.EPHEMERON_TOMBSTONE ? (e.key, e.value) : default;
```

Reply:
I have to say that I asked myself the same exact question as well while refactoring 😅
I left it as it was for now just to avoid making the PR more complex than intended.
This is how the type is currently implemented in Mono:

https://github.com/dotnet/runtime/blob/ea69731ad1447345a01a4d918898dea8456bda68/src/mono/System.Private.CoreLib/src/System/Runtime/CompilerServices/DependentHandle.cs#L45-L51

Which to me does look like you could end up reading a tombstone and a null if the GC stopped everything right after the branch and collected the key and value. But then again I'd have expected such an issue to have come up already given that `ConditionalWeakTable<,>` is heavily used already? 🤔

Let me know if you'd still like me to change these implementations on Mono anyway!

Reply:
Yes, it would be nice to fix.

>  I'd have expected such an issue to have come up already given that ConditionalWeakTable<,> is heavily used already?

It is a race condition. Race conditions like this can linger in the product for a long time. The GC has to happen at the right spot, there needs to be a follow up code that leads to a crash, and we must be lucky to capture the crashdump and connect the dots to this bug. Also, Mono is typically used in less stressful workloads vs. CoreCLR so it makes it even less likely for this bug to be found and fixed.

Reply:
I see, thanks! I got that it needed some very specific conditions to happen and to also be properly detected, I guess I just thought Mono was being used so much considering all the various platforms it supports, and `CWT<,>` being pretty common lower in the stack, that it'd have caused more noticeable issues already 😄

Fixed in 9f7b212d39f6d9b39db764a3032f0a4ff7dddf38 for all APIs.

Reply:
> `CWT<,>` being pretty common lower in the stack

In dotnet/runtime, `CWT<,>` is only used on paths that are rarely used or generally slow for other reasons. It is not stressed much by the lower levels of the stack.

Reply:
Right, I was thinking that given that eg. `ArrayPool<T>.Shared` uses it too now ([here](https://github.com/dotnet/runtime/blob/a02b668214e9711cb3d5f426f085601698a24b92/src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs#L56-L57)) and that type is getting relatively used more and more frequently, that could influence that, but as you said that's only really used in slower paths, so probably not under all that actual stress after all. I guess really I'm just always surprised whenever bugs like this are discovered in the runtime - of course no software is actually bug-free, but one often tends to think of the runtime as being almost infallible 😄